buildstate-maven-extension/src/main/java/io/helidon/build/state/maven/ProjectStateLifecyclePartitipant.java

```diff
@@ -0,0 +1,243 @@
+    private LifecycleMappingDelegate standardLifecycleDelegate;
+
+    @Requirement
+    private Map<String, LifecycleMappingDelegate> lifecycleDelegates;
+
+    @Requirement
+    private LifecycleExecutionPlanCalculator executionPlanCalculator;
+
+    private Logger logger;
+
+    @Override
+    public void enableLogging(Logger logger) {
+        this.logger = logger;
+    }
+
+    @Override
+    public void afterProjectsRead(MavenSession session) throws MavenExecutionException {
+        if (session.getGoals().contains("clean")) {
+            logger.debug("clean requested, project state is ignored");
+        } else {
+            for (MavenProject project : session.getProjects()) {
+                try {
+                    ProjectState state = ProjectState.load(project);
+                    if (state == null) {
+                        logger.debug("Project state file not found");
+                        continue;
+                    }
+                    if (!state.projectFiles().equals(ProjectFiles.of(project))) {
+                        logger.debug("Project files changed: [" + project.getGroupId() + ":" + project.getArtifactId() + "]");
+                        continue;
+                    }
+                    state.apply(project);
+
+                    // resolve the executions in the current life-cycle and filter only the ones that were recorded
+                    List<RecordedExecution> executions = resolveExecutions(session, project, state);
+
+                    // skip the recorded executions
+                    project.getBuild().getPlugins()
+                           .forEach(plugin -> skipExecutions(plugin, executions));
+
+                    projectStates.put(project, state);
+                } catch (IOException ex) {
+                    logger.error("Error while loading project state", ex);
+                }
+            }
+        }
+
+        Map<MavenProject, List<RecordedExecution>> executions = new HashMap<>();
+        MavenExecutionRequest request = session.getRequest();
+        request.setExecutionListener(new DelegatingExecutionListener(request.getExecutionListener()) {
+
+            @Override
+            public void mojoSucceeded(ExecutionEvent event) {
+                MojoExecution execution = event.getMojoExecution();
+                if ("default-cli".equals(execution.getExecutionId())) {
+                    return;
+                }
+                executions.computeIfAbsent(event.getProject(), p -> new LinkedList<>())
+                          .add(RecordedExecution.create(execution));
+                super.mojoSucceeded(event);
+            }
+
+            @Override
+            public void projectSucceeded(ExecutionEvent event) {
+                MavenProject project = event.getProject();
+                try {
+                    ProjectState.merge(projectStates.get(project), project, executions.getOrDefault(project, List.of()))
+                                .save(project);
```

Comment:
Isn't it possible that concurrent builds will step on each other here?

Reply:
do you mean concurrent maven processes building the same module? If so there isn't much we can do. Do you have something in mind ?

Reply:
Yes. You should get an exclusive lock on the file for all accesses in `ProjectState`.